3 better compiler support

hiker opened the pull request:
This is the big one - it introduces `Tool` as baseclass for anything called via subprocess (and removed `run_command` therefore). It introduces the ToolRepository and  ToolBox. Best to read the documentation (atm site-specific-configuration.rst).

Note that this is not really complete, outstanding issues are support for compiler-specific flags, better support for linking, handling openmp (so that we can trigger fparser's OMP sentinel handling instead of using a work around).

We did fix up the run_configs, but they won't really work better than before (as in, still hard-coded compiler etc). Once we have a fully working LFRic build system (including the above todos), we will update all the run configs.

MatthewHambley commented on tests/unit_tests/test_config.py:
```diff
@@ -1,15 +1,16 @@
+from fab.tools import ToolBox
 
 
 class TestAddFlags(object):
```
```suggestion
class TestAddFlags:
```

MatthewHambley commented on tests/unit_tests/tools/test_versioning.py:
Maybe take advantage of test classes to group tests for different version control systems.

MatthewHambley reviewed (changes requested):
This change represents a step in the right direction and should be enabling for a lot of good future work.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+:class:`~fab.tools.tool_box.ToolBox` and pass
+it to the :class:`~fab.build_config.BuildConfig` object. This toolbox
+contains all the tools that will be used during the build process, but
+it can only store one tool per category. If a certain tool should not
```
There are two approaches:
1. (which is what we use) We create `Meta` compiler class: its constructor takes two existing compilers instances (e.g. normal compiler and profile-wrapper), and its compile method will decide which compiler to use (e.g. based on path, which is what we use, we exclude all `.../kernels/...` from being instrumented, since they are called way too frequent). This could even be based on a static/dynamic dictionary being defined (similar to path flags)
2. The user could call the compile step more than once (with different source getter), and modify the toolbox in between. The problem with this is that it would not take dependencies into account, i.e. can all files be compiled with one compiler before the other??

I think the use case of this is reasonable infrequent that the first approach is good enough (esp. since you could write a constructor for a meta compiler that specifies which compiler to use for which path)

hiker replied:
My wording was perhaps a bit bad ... meta-compiler does not clearly describe what I meant :)   I totally agree, wrapped compilers are VERY important. I was meant to focus on 'specific compiler for specific files' (e.g. what we need for profiling, where we don't want to instrument kernels) are being rare. Seeing that we agree in #312, I am closing this thread then.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+A site might have compilers that Fab doesn't know about, or prefers
+a different compiler from the Fab default. Fab abstracts the compilers
+and other programs required during building as an instance of a
+:class:`~fab.tools.tool.Tool` class. All tools that Fab knows about, are
```
Indeed, I forgot. one `tool` is not necessary (since evertything is imported in `tools/__init__.py`). I've fixed these occurrences.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+====
+Each tool must be derived from :class:`~fab.tools.tool.Tool`.
+The base class provides a `run` method, which any tool can
+use to execute a command in a shell. Typically, a tool will
```
You can easily overwrite the run function of the `Tool` base class and just call a Python function. I actually considered doing this when I noticed the templaterator, but decided against it since I would assume that the CLI will stay identical, while the implementation might change(?)

hiker replied:
Agreed.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+==========
+All possible categories are defined in
+:class:`~fab.tool.categories.Categories`. If additional categories
+should be required, they can be added.
```
That needs to modify the sources, but see #311, where I suggest to generalise the usage of `MISC`, meaning user can easily add tools to the `MISC` category.

Alternative, we could allow `Union[str, Categories]`, then a user can add arbitrary strings, and it would certainly not interfere with the Fab-hard-coded categories.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+For example, a compiler instance provides a
+:func:`~fab.tools.compiler.Compiler.compile_file` method.
+This makes sure that no tool-specific command line options need
+to be used in any Fab step, which will allow to replace any tool
```
Done.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+created by providing a different set of parameters in the
+constructor.
+
+This also allows to easily define compiler wrapper. For example,
```
Done.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+constructor.
+
+This also allows to easily define compiler wrapper. For example,
+if you want to use `mpif90` as compiler, which is a MPI-specific
+wrapper for `ifort`, you can create this class as follows:
```
Well, the difficulty is that even though the wrapper is always called `mpif90` (or whatever), but it will need different options to e.g. specify module output path, openmp etc.

But I will need to think about the linker handling, e.g. supporting non-MPI builds?? Do we need that? Obviously, we can use any MPI compiler wrapper for non-MPI programs without much harm (minor link time increase since it might search the MPI libraries??)

hiker replied:
Done.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+
+    from fab.tools import Ifort
+
+    class MpiF90(Ifort):
+        '''A simple compiler wrapper'''
+        def __init__(self):
+            super().__init__(name="mpif90-intel",
+                             exec_name="mpif90")
```
Yes, except for vendor-specific flags (module path, openmp, ...).

hiker replied:
Yes, I believe we agree here, see #312.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+:func:`~fab.tools.tool_repository.ToolRepository.set_default_vendor`
+which allows you to change the defaults for compiler and linker with
+a single call. This will allow you to easily switch from one compiler
+to another.
```
Yes, in the tool box you can (if you want) start with setting a vendor default, then overwrite each individual compiler with whatever you want. I've added this to the text, thanks!

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+Tool Box
+========
+The class :class:`~fab.tools.tool_box.ToolBox` is used to provide
+the tools to be use to the build environment, i.e. the
```
Done.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+========
+The class :class:`~fab.tools.tool_box.ToolBox` is used to provide
+the tools to be use to the build environment, i.e. the
+BuildConfig object:
```
Done.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+
+The advantage of finding the compilers to use in the tool box is that
+it allows a site to replace a compiler in the tool repository (e.g.
+if a site wants to use an older gfortran version, say one which is called
```
I've added some explanation here.

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -0,0 +1,142 @@
+the default tool from the `ToolRepository` will be used. Therefore,
+in the example above adding `ifort` is not strictly necessary (since
+it will be the default after setting the default vendor to `intel`),
+and `c_comp` is the default as well. This feature is especially useful
```
Done.

hiker commented on run_configs/build_all.py:
```diff
@@ -4,20 +4,33 @@
 def build_all():
 
+    tool_box = ToolBox()
+    compiler = tool_box[Categories.FORTRAN_COMPILER]
```
I would tend to agree (and am happy to read your opinion :) ). I expect a lot of the examples to be changed later (e.g. our build scripts for LFRic look quite different from what is currently in run_configs), but since we are changing quite a lot, my idea is that we update all run_config once we done with the changes (e.g. better compiler-specific flags, and handling of linker). We will maintain the scripts so that they kind of work the way they worked originally till then.

hiker commented on run_configs/build_all.py:
Agreed.

hiker commented on run_configs/jules/build_jules.py:
```diff
@@ -13,26 +13,37 @@
 logger = logging.getLogger('fab')
 
 
+class MpiIfort(Ifort):
+    '''A small wrapper to make mpif90 available.'''
+    def __init__(self):
+        super().__init__(name="mpif90", exec_name="mpif90")
+
+
```
Indeed, that is my current thinking, but I am getting second thoughts ... I think it might be better if we used a Decorator pattern, that might make it much easier to handle compiler-specific flags (which after all also needs to be used in linking as well, e.g. omp flags etc).

But I think this will just change the way we implement compiler-wrapper (mpif90 around a standard compiler) and linker (which I think should pretty much be a compiler with the specified compiler flags, but additionally adds linker-flags for libs etc). It should not affect anything implemented in this PR, and we will verify what is best once we start working on compiler flags.

hiker commented on run_configs/jules/build_jules.py:
```diff
@@ -13,26 +13,37 @@
 logger = logging.getLogger('fab')
 
 
+class MpiIfort(Ifort):
```
Dang, good point. I believe the original code had `mpifort`, and I jumped to the wrong conclusion (`mpiifort` :) ), so I replaced it with Ifort :(

For now, I've added a todo #312 to build_jules.py, though it will affect all examples there, and do a proper implementation there, I hope that's ok.

hiker commented on run_configs/jules/build_jules.py:
```diff
@@ -13,26 +13,37 @@
-    with BuildConfig(project_label=f'jules {revision} $compiler') as state:
+    tool_box = ToolBox()
+    # Create a new Fortran compiler MpiIfort
+    fc = MpiIfort()
```
Not sure if I understand you correctly here. Internally, in BuildConfig, `$compiler` will be replaced with the name of the Fortran compiler in the tool box. So the ToolBox defines `$compiler` (not the other way round??)

hiker commented on run_configs/jules/build_jules.py:
```diff
@@ -13,26 +13,37 @@
         fcm_export(state, src='fcm:jules.xm_tr/utils', revision=revision, dst_label='utils')
 
-        #
-        grab_pre_build(state, path='/not/a/real/folder', allow_fail=True),
```
I thought that as not supposed to be there :) I've added it back in.

hiker commented on run_configs/tiny_fortran/build_tiny_fortran.py:
```diff
@@ -4,19 +4,36 @@
+from fab.tools import Ifort, Linker, ToolBox
+
+
+class MpiIfort(Ifort):
```
Yes, indeed, my mistake :(  See #312. I would prefer to have the proper solution as stand-alone PR then just hacking all these scripts now to test for gfortran and ifort ;(

hiker commented on source/fab/artefacts.py:
```diff
@@ -20,6 +20,21 @@
 from fab.util import suffix_filter
 
 
+class ArtefactStore(dict):
```
Yes. TBH, I think ideally that would become its own subdirectory, with each class its own file (which is what we try to get to in PSyclone, though of course some classes are just too tiny :( ). I just left it all in one file to keep this PR smaller.

hiker commented on source/fab/cli.py:
```diff
@@ -16,69 +19,56 @@
+    # TODO: This assumes a mapping of compiler name to the corresponding
+    # linker name (i.e. `linker-gfortran` or `linker-ifort`). Still, that's
+    # better than hard-coding gnu here.
+    linker = tr.get_tool(Categories.LINKER, f"linker-{fc.name}")
```
Indeed.

hiker commented on source/fab/parse/__init__.py:
```diff
@@ -39,11 +39,11 @@ def __init__(self, fpath: Union[str, Path], file_hash: Optional[int] = None):
 
     @property
-    def file_hash(self):
+    def file_hash(self) -> int:
```
That would indeed be neat.

hiker commented on source/fab/steps/compile_c.py:
```diff
@@ -32,11 +30,10 @@
 
 @dataclass
-class MpCommonArgs(object):
+class MpCommonArgs():
```
Oops.

hiker commented on source/fab/steps/compile_c.py:
```diff
@@ -66,18 +62,11 @@ def compile_c(config, common_flags: Optional[List[str]] = None,
-    compiler_version = get_compiler_version(compiler)
-    logger.info(f'c compiler is {compiler} {compiler_version}')
+    compiler = config.tool_box[Categories.C_COMPILER]
+    logger.info(f'c compiler is {compiler}')
```
Done.

hiker commented on source/fab/steps/compile_c.py:
```diff
@@ -121,27 +110,24 @@ def store_artefacts(compiled_files: List[CompiledFile], build_lists: Dict[str, L
     with Timer() as timer:
-        flags = mp_payload.flags.flags_for_path(path=analysed_file.fpath, config=mp_payload.config)
-        obj_combo_hash = _get_obj_combo_hash(mp_payload.compiler, mp_payload.compiler_version, analysed_file, flags)
+        flags = Flags(mp_payload.flags.flags_for_path(path=analysed_file.fpath,
```
I can't agree more. It's our next big work package :)

hiker commented on source/fab/steps/compile_c.py:
```diff
@@ -152,14 +138,13 @@ def _compile_file(arg: Tuple[AnalysedC, MpCommonArgs]):
-            flags_checksum(flags),
-            zlib.crc32(compiler.encode()),
-            zlib.crc32(compiler_version.encode()),
+            flags.checksum(),
+            compiler.get_hash(),
```
Good.

hiker commented on source/fab/steps/compile_fortran.py:
```diff
@@ -8,44 +8,37 @@
 
 """
 
-# TODO: This has become too complicated. Refactor.
```
OK.

hiker commented on source/fab/steps/compile_fortran.py:
```diff
@@ -122,32 +114,19 @@ def compile_fortran(config: BuildConfig, common_flags: Optional[List[str]] = Non
-    compiler_version = get_compiler_version(compiler)
-    logger.info(f'fortran compiler is {compiler} {compiler_version}')
+    compiler = config.tool_box[Categories.FORTRAN_COMPILER]
+    logger.info(f'fortran compiler is {compiler} {compiler.get_version()}')
```
Done.

hiker commented on source/fab/steps/compile_fortran.py:
```diff
@@ -8,44 +8,37 @@
 
 @dataclass
-class MpCommonArgs(object):
+class MpCommonArgs():
```
Oops again.###

hiker commented on source/fab/steps/link.py:
```diff
@@ -63,31 +45,29 @@ def link_exe(config, linker: Optional[str] = None, flags=None, source: Optional[
-    linker = linker or os.getenv('LD', 'ld')
-    logger.info(f'linker is {linker}')
+    linker = config.tool_box[Categories.LINKER]
+    logger.info(f'linker is {linker.name}')
```
Done.

hiker commented on source/fab/steps/psyclone.py:
Partially agreed. At this stage, all Fab build examples we have looked at or will be looking at in the near future involve PSyclone (LFRic, UM ... in NG-Arch the UM physics will be using PSyclone, a Tsunami code we are doing some work with, NEMO). So it is really convenient to have this provided as part of Fab.

hiker replied:
Can't agree more ;) We will look at that in a bit more detail ;) After all, it all depends on which steps are called, and with an OO design we can easily have a base class without PSyclone, but add this (as default) for LFRic etc.

hiker commented on source/fab/steps/psyclone.py:
No, see line 299:
```
        psyclone = config.tool_box[Categories.PSYCLONE]
```

hiker commented on source/fab/tools/ar.py:
```diff
@@ -0,0 +1,46 @@
+            requesting the ar version.
+        '''
+        try:
+            self.run("--version")
```
Hmm - not sure which one is better. Could we have an executable in the path that is not executable? While certainly not applicable for any useful example here, what if we have a dynamic binary without RPATH and incorrect LD_LIBRARY_PATH? That would probably still report under `which`, but not be executable?
``` 
~/fab-workspace/skeleton-mpif90$ which ./skeleton
./skeleton
:~/fab-workspace/skeleton-mpif90$ ./skeleton 
./skeleton: error while loading shared libraries: libyaxt.so.1: cannot open shared object file: No such file or directory
```

But I am happy to change, I don't think it will make a difference in reality.

hiker commented on source/fab/tools/categories.py:
```diff
@@ -0,0 +1,37 @@
+from enum import auto, Enum
+
+
+class Categories(Enum):
```
See #311.

hiker commented on source/fab/tools/categories.py:
```diff
@@ -0,0 +1,37 @@
+        return str(self.name)
+
+    @property
+    def is_compiler(self):
```
Because the ToolRepository adds automatically each compiler as linker. While this could just check the Tool's Categories of course, it would be limited to compilers that are specified in the Categories, while this way a user *could* add a new compiler as MISC tool, and would still get a linker for that new language.

But tbh, I don't see this as a real use case ;)

I might change this as part of #312. One option could also be to rename this function to `can_be_used_as_linker`?

hiker commented on source/fab/tools/compiler.py:
```diff
@@ -0,0 +1,301 @@
+    def get_hash(self) -> int:
+        ''':returns: a hash based on the compiler name and version.
+        '''
+        return (zlib.crc32(self.name.encode()) +
```
I don't think we should hash() - afaik the python has values are randomly (i.e. 'salted' at startup with a random offset, to avoid a potential DOS attach by forcing hash values clusters), which would mean that the hash for anything would change from run to run.

MatthewHambley replied:
I've raised ticket #317 to investigate this.

hiker replied:
I've answered :)

hiker commented on source/fab/tools/compiler.py:
```diff
@@ -0,0 +1,301 @@
+            this by requesting the compiler version.
+        '''
+        try:
+            version = self.get_version()
```
Let me know if you prefer a `which` - it certainly means that we could move more of the very similar code from the various tools into the base class. I think I stuck with a help/version command, since this was used for some tools in the past, so that's how I started :) But I don't mind one way or another.

hiker replied:
Ah, I forgot one argument that swayed me: since I expect scripts might want to check compiler versions at some stage, using the compiler version as a test will save one call :)    (admittedly, it doesn't matter)

hiker commented on source/fab/tools/compiler.py:
```diff
@@ -0,0 +1,301 @@
+            return self._version
+
+        try:
+            res = self.run("--version", capture_output=True)
```
Agreed. From memory, a Cray compiler might have to overwrite this function. 

OK, the only argument for using a --verison/--help/.. option is the admittedly stupid case of a binary that's not runnable (due to ld_library_path or so). Let me know what you prefer, and I'll update this.

hiker replied:
When adding compiler wrapper, I will also add Cray's I'd guess, so that will be properly handled.

hiker commented on source/fab/tools/compiler.py:
```diff
@@ -0,0 +1,301 @@
+
+        # Pull the version string from the command output.
+        # All the versions of gfortran and ifort we've tried follow the
+        # same pattern, it's after a ")".
```
And that's the reason why there is no Cray compiler defined atm :)  My idea is that we properly handle version (i.e. breaking them into integer-tuples, storing them, and using them in version-specific tests). At that time I would add Cray plus a custom function to query and extract the version number.

MatthewHambley replied:
Great. Please raise a ticket with the details so it doesn't get lost.

hiker replied:
#318. Note that it will be a while before we get to that, since we have to support compiler-version-specific compiler flags, so changes to the flag handling need to come first (or well, all together ... or maybe I split this in the end ... anyway, work-in-progress).

hiker commented on source/fab/tools/compiler.py:
```diff
@@ -0,0 +1,301 @@
+        self._module_output_path = ""
+        self._syntax_only_flag = syntax_only_flag
+
+    @property
+    def has_syntax_only(self) -> bool:
+        ''':returns: whether this compiler supports a syntax-only feature.'''
+        return self._syntax_only_flag is not None
+
+    def set_module_output_path(self, path: Path):
+        '''Sets the output path for modules.
+
+        :params path: the path to the output directory.
+        '''
+        self._module_output_path = str(path)
```
Oh, I am not uo-to-date with latest C++ :(  Haven't done any C++ for ... don't know, over 7 years or so. BOM is a Fortran and python shop mostly.

hiker commented on source/fab/tools/flags.py:
```diff
@@ -0,0 +1,78 @@
+        '''
+        i = 0
+        flag_len = len(remove_flag)
+        while i < len(self):
```
I already have a ticket for flag improvements - https://github.com/hiker/fab/issues/22 . Esp. fixes, since this PR badly introduces a new flag object ... my apologies, when I saw how big this PR became, I left the flag pretty basic, to be fixed later). For improved visibility, I've opened #313 (which also references our internal ticket)

hiker commented on source/fab/tools/psyclone.py:
```diff
@@ -0,0 +1,78 @@
+from fab.tools.categories import Categories
+from fab.tools.tool import Tool
+
+if TYPE_CHECKING:
```
I can't agree more, I've really tried to avoid this, but both classes needs this for typing, so I couldn't use the usual easy solution of just locally importing the symbol :( See #314 .

hiker commented on source/fab/tools/rsync.py:
```diff
@@ -0,0 +1,51 @@
+# which you should have received as part of this distribution
+##############################################################################
+
+"""This file contains the Rsync class for archiving files.
```
That's indeed much better.

hiker commented on source/fab/tools/tool.py:
```diff
@@ -0,0 +1,192 @@
+        return self._is_available
+
+    @is_available.setter
+    def is_available(self, value: bool):
```
No, you are totally right. I needed it for some testing, but I just changed the tests to set the private attribute directly.

hiker commented on source/fab/tools/tool.py:
```diff
@@ -0,0 +1,192 @@
+        ''':returns: the flags to be used with this tool.'''
+        return self._flags
+
+    @property
+    def logger(self) -> logging.Logger:
+        ''':returns: a logger object for convenience.'''
+        return self._logger
```
Yeah, not sure here. I tried to avoid the repeated import of logging etc (I really don't like code execution on import time, so all these logging calls triggered me :) ). I am not sure this is better, though it does mean none of the tools need to import/setup logging.

hiker commented on source/fab/tools/tool_box.py:
```diff
@@ -0,0 +1,53 @@
+        '''
+        if not tool.is_available:
+            raise RuntimeError(f"Tool '{tool}' is not available.")
+        self._all_tools[tool.category] = tool
```
That's done on purpose. I want to support that we eg. inheriting a setup, and do minimal changes (e.g. for our profiling, we might use a fab setup that compiles with ifort, but we would just replace ifort with our Meta-compiler-wrapper).

MatthewHambley replied:
I see your use case, I just worry that it may not be appropriate default behaviour. What do you think of a "silently replace" flag to the function?

hiker replied:
Done.

hiker commented on source/fab/tools/tool_box.py:
```diff
@@ -0,0 +1,53 @@
+    '''
+
+    def __init__(self):
+        self._all_tools = {}
```
Good point, I still need to get my ahead around typing. BUT, when I also add the return type on `__getitem__`:
```
    def __getitem__(self, category: Categories) -> Tool:
```
I suddenly get 8 errors (that are no problem without this additional and unnecessary?? type information):
```
source/fab/steps/preprocess.py:158: error: Argument "preprocessor" to "pre_processor" has incompatible type "Tool"; expected "Preprocessor"  [arg-type]
```
All of them are kind of right, it gets a `Tool`, but expects a `Preprocessor`. What should I do?
1. Remove the return type of `__getitem`?
2. Use a type case everywhere to fix these errors?
3. Change the specific tool type in each case from (say) Preprocessor to Tool (though I expect I then get problems with  calling pre-processor specific functions in a 'tool'???  And I gave each tool a different `do_something` method (e.g. compile, link), since pylint does not like to overwrite a function (`run`) with one that takes different parameters - plus it's a lot more descriptive what is happening).

:(

MatthewHambley replied:
If a routine needs specifically a `Preprocessor` rather than a generic `Tool` then it should specify that in the argument list. It then becomes the caller's responsibility to make sure it is passing one. e.g.

```
if not isinstance(tool, Preprocessor):
    raise TypeError("Need a preprocessor")
thing.method(tool)
```

Alternatively you could do the check in the function

```
def method(self, tool: Tool):
    assert isinstance(tool, Preprocessor)
```

The first has the advantage of providing more static type checking but it is a bit ugly. The second is cleaner but reduces static type checking.

hiker replied:
Ah, that is great, I had no idea that mypy was that smart. I've done it the first way, if we go through all the effort of adding types, we should use them as much as possible.

hiker commented on source/fab/tools/tool_box.py:
```diff
@@ -0,0 +1,53 @@
+            raise RuntimeError(f"Tool '{tool}' is not available.")
+        self._all_tools[tool.category] = tool
+
+    def get_tool(self, category: Categories):
```
(actually, the comment for line 21 should probably better be here ;) )

hiker commented on source/fab/tools/tool_repository.py:
```diff
@@ -0,0 +1,142 @@
+        # we only want to initialise the instance once, so let the constructor
+        # not do anything if the singleton already exists:
+        if ToolRepository._singleton:
+            return
```
The only other way I am aware of is using (e,g.) `ToolRepository.get()` - which I had originally :)  And I can add a test in the constructor that the singleton instance is still None (or maybe set a custom value to guarantee the constructor is only called from the static singleton getter)?

MatthewHambley replied:
Singleton seems to be a pattern which depends on data hiding for slick function. Something which Python deliberately eschews. Lets leave it as it is for the moment and see how we feel about it later.

hiker replied:
I am aware that in general Singleton in Python is disliked/not recommended, so I am happy to discuss this later. FWIW, I like singleton because it gives a single function (the constructor)which can be used to setup an instance. I might be missing something, but without this we would have to have a test in any function to make sure the class/object/module/package (whatever) is correctly set up.

hiker commented on source/fab/tools/tool_repository.py:
```diff
@@ -0,0 +1,142 @@
+        raise KeyError(f"Unknown tool '{name}' in category '{category}' "
+                       f"in ToolRepository.")
+
+    def set_default_vendor(self, vendor: str):
```
I am happy to have a look/discussion. FWIW, whatr about renaming 'vendor' to 'compiler_suite` ? Then we could have `intel-classic`, and `intel-llvm` (plus `gnu`, `cray`, `nvidia` I'd guess).
Given that otherwise vendor (well. compiler_suite) is a string, any user can add any add whatever they want.

Re other tools - is there a need to have a vendor? If you want (say) specifically a gnu-tar, gnu-make, you can always just create a tool class correspondingly (i.e. either with full path, or with the proper name like `gmake`).

hiker commented on source/fab/tools/versioning.py:
```diff
@@ -0,0 +1,241 @@
+            return False
+        return True
+
+    def is_working_copy(self, path: Union[str, Path]) -> bool:
+        """:returns: whether the given path is a working copy or not. It
+            runs the command specific to the instance.
+
+        :param path: directory to be checked.
+        """
+        try:
+            self.run([self._working_copy_command], cwd=path,
+                     capture_output=False)
+        except RuntimeError:
+            return False
+        return True
```
It is used in `steps/grab/{git,svn}` - and I am pretty sure I implemented it because these functions used it in the first place. I am happy to remove this (which means the grab steps will likely fail with an exception indicating that this is not  a working directory, which I am totally ok with :) )

MatthewHambley replied:
Lets have it out and reduce this size of this change by a few lines.

hiker replied:
Really? I am surprised :) OK, I have done it in a stand-alone commit so we can undo it in case that I misunderstood. So the svn and git steps do not check anymore if they are running on a correct repository anymore, instead this will now raise a RuntimeError (by svn/git).

hiker commented on source/fab/util.py:
```diff
@@ -110,12 +110,12 @@ def file_walk(path: Union[str, Path], ignore_folders: Optional[List[Path]] = Non
 
 
-class Timer(object):
+class Timer():
```
Done.

hiker commented on tests/system_tests/git/test_git.py:
```diff
@@ -23,12 +23,13 @@
+    return BuildConfig('proj', ToolBox(), fab_workspace=tmp_path)
 
 
 class TestGitCheckout(object):
```
Done.

hiker commented on tests/system_tests/psyclone/test_psyclone_system_test.py:
```diff
@@ -61,7 +63,7 @@ def test_make_parsable_x90(tmp_path):
 
 
-class TestX90Analyser(object):
+class TestX90Analyser():
```
Done.

hiker commented on tests/unit_tests/steps/test_root_inc_files.py:
```diff
@@ -5,15 +5,16 @@
 
 
-class TestRootIncFiles(object):
+class TestRootIncFiles():
```
Done.

hiker commented on tests/system_tests/svn_fcm/test_svn_fcm_system_test.py:
```diff
@@ -26,12 +28,14 @@
 
-if tool_available('svn'):
+svn = Subversion()
+if svn.is_available:
     export_funcs.append(svn_export)
     checkout_funcs.append(svn_checkout)
     merge_funcs.append(svn_merge)
 
-if tool_available('fcm'):
+fcm = Fcm()
+if fcm.is_available:
     export_funcs.append(fcm_export)
     checkout_funcs.append(fcm_checkout)
     merge_funcs.append(fcm_merge)
```
Yes, that would be niced (fwiw, I did just leave it the way it was :) )

hiker commented:
Thanks a lot for the review. I believe I have addressed most issues. Outstanding:
1. I am not sure about fixing the typing - if I fix it, I need to use quite a few casts. I think that might be ok, but if you have a better idea please let me know.
2. Proper handling of compiler wrapper. I started to fix it up here, but I think that becomes to big and should better be done standalone later.
3. Improved linker handling - it's still a bit odd.
4. I need some feedback re adding new (user defined) tools (#311): should I merge all tool categories into a generic 'misc' one, which will make it a lot easier for the user to add new tools? To me that feels sufficient (though that means that many generic tools like rsync, svn, git, ... will be found by name, instead of the corresponding category. I think that's fine, since even if a site has a different named git version, we can still just add this as a separate tool - using the 'standard' name (say git), but calling git-3.1415 as executable name. Just remove the old git, and replace it with a new one.
5. Default intel compiler in run_configs: while I could fix this now, it would lead to a lot of code duplication. I propose to leave this for now, once we have all other things fixed, I believe it will be a lot less code duplication (e.g. if the ToolRepository can check to automatically find the right mpif90 version of a specified compiler etc).
6. I've lost the comment atm - while I did not use `which` for tool availability (since using a specific command has the advantage that it tests things like ld_library_path), I moved this handling into the base class, which removes a lot of duplicate code (and still allows to customise the command to use per tool). I did use `which` in one of the run_configs example.

I believe there might be another issue or two that I should mentioned here, but I believe I have comments on all your feedback, so it will be in mu answers. I've opened a few tickets (#311, #312, #313, #314) to keep track of follow-up work.

Thanks for the review

MatthewHambley commented on source/fab/tools/tool.py:
```diff
@@ -28,17 +27,25 @@ class Tool:
         self._logger = logging.getLogger(__name__)
         self._name = name
-        self._exec_name = exec_name
+        self._exec_name = str(exec_name)
```
At the moment this is fine since the executable filename is only ever used to compile the argument list, which is a list of strings. If we ever need to modify the path or perform tests on it we may prefer to keep it as a `Path`. e.g. `self._exec_name.is_executable()`.

hiker replied:
OK. It would be easy to push a path through if needed.

MatthewHambley reviewed (changes requested):
Good progress, I think we are close on this change.

Please conclude any conversations you are happy with to let me know that they are done with.

hiker commented on run_configs/build_all.py:
```diff
@@ -4,20 +4,33 @@
+                         category=Categories.MISC)
+
+    def check_available(self):
+        return True
```
I've updated this using `shutil.which`, afaik there is no is_executable in pathlib objects.

hiker commented on source/fab/tools/tool_repository.py:
```diff
@@ -0,0 +1,142 @@
+
+        # Create the list that stores all tools for each category:
+        for category in Categories:
+            self[category] = []
```
I forgot to answer this apparently: I can't inherit from a `defaultdict(list)`, so not sure how to change this (short of not making dictionary a base class, but that might then mean we need to re-implement some of the required dictionary functions, which we currently get for free?

hiker commented on docs/source/site-specific-config.rst:
```diff
@@ -80,32 +88,35 @@ startup section can add more tools to the repository:
+Compiler and linker objects define a compiler suite, and the `ToolRepository`
 provides
-:func:`~fab.tools.tool_repository.ToolRepository.set_default_vendor`
+:func:`~fab.tools.tool_repository.ToolRepository.set_default_compiler_suite`
```
Nothing that I can think of atm. I think tbh even the flexibility we offer for the other tools will de-facto not be used (anyone on solaris? That apparently needs gtar from what I remember :) ).
I am happy to leave this for now, and if as part of #311 this changes, we can still rename.

hiker commented:
OK, I believe I have addressed all issues, and as you suggested I closed all conversations where I think we have agreed :) I left a few things open where I think you should provide feedback to confirm :)

MatthewHambley reviewed (approved):
Time to take the plunge.
